massa-models/src/block.rs

```diff
@@ -759,28 +835,54 @@ impl Deserializer<BlockHeader> for BlockHeaderDeserializer {
             parents[slot.thread as usize],
         ));
 
-        let (rest, endorsements) = context(
+        let (rest, endorsements): (&[u8], Vec<Wrapped<Endorsement, EndorsementId>>) = context(
             "Failed endorsements deserialization",
             length_count::<&[u8], Wrapped<Endorsement, EndorsementId>, u32, E, _, _>(
                 context("Failed length deserialization", |input| {
                     self.length_endorsements_deserializer.deserialize(input)
                 }),
                 context("Failed endorsement deserialization", |input| {
-                    endorsement_deserializer.deserialize_with(&self.endorsement_serializer, input)
+                    let (rest, endo) = endorsement_deserializer
+                        .deserialize_with(&self.endorsement_serializer, input)?;
+
+                    if endo.verify_signature().is_err() {
+                        Err(nom::Err::Failure(ContextError::add_context(
+                            rest,
+                            "Endorsement signature invalid",
+                            ParseError::from_error_kind(rest, nom::error::ErrorKind::Fail),
+                        )))
+                    } else {
+                        Ok((rest, endo))
+                    }
                 }),
             ),
         )
         .parse(rest)?;
 
-        Ok((
-            rest,
-            BlockHeader {
-                slot,
-                parents,
-                operation_merkle_root,
-                endorsements,
-            },
-        ))
+        let parent_id = parents[slot.thread as usize];
+        for endo in endorsements.iter() {
+            if endo.content.endorsed_block != parent_id {
+                return Err(nom::Err::Failure(ContextError::add_context(
+                    rest,
+                    "Endorsement does not match block parents",
+                    ParseError::from_error_kind(rest, nom::error::ErrorKind::Fail),
+                )));
+            }
+        }
+
+        let header = BlockHeader {
+            slot,
+            parents,
+            operation_merkle_root,
+            endorsements,
+        };
+
+        {
+            #[cfg(any(test, feature = "testing"))]
+            assert!(header.assert_invariants().is_ok());
+        }
```

Comment:
Thinking behind this, is that asserting invariants within this function makes for much simpler code, and is less prone to errors/bugs. By making this available in any environment run with the testing flag, it tests more thoroughly for errors in the deserialiser. 

There are the obvious drawbacks of code that is somewhat redundant.

shall we keep it around?